fusion/README.md

```diff
@@ -0,0 +1,198 @@
+the shim or increase the number of FieldExtractor setter/getter methods.
+
+There are multiple input mechanisms intended to simplify implementing
+the fusion operation. Currently I am using UID. There will be support
+for opcodes, assembly text, a DSL syntax and possibly a Json input format.
+
+UID based fusion is the example in this draft. The others are planned.
+
+## Other classes
+
+The support classes for Fusion are:
+
+```
+  FusionContext         fusioncontext.h
+  FusionExceptionBase   fusionexceptions.h
+  using and typedefs    fusiontypes.h
+  RadixTrie             radixtrie.h
+```
+
+There is a single context supported in this version of the implementation but
+generalization to multiple contexts has placeholders in the source.
+
+FusionExceptionBase is the parent for a a number of fusion/dsl/json/run-time exceptions specific to fusion.
+
+RadixTrie is experimental. It is implemented and provided for performance analysis during development. The currently used container is an unordered\_map 
+used for functionality and test development.
+
+## Usage within a Sparta unit
+
+The currently considered procedure for Fusion is an implementation in a 
+Sparta unit typically in the Decode. The primary input to the 
+fusionOperator() is a form of the InstQueue, sparta::Queue<InstPtr>.
+(Other forms such as InstGroup are likely).  An output buffer of similar type is also provided to fusionOperator(). The fusionOperator() performs the
+search across the current FusionGroup context and executes the transform
+functor in the FusionGroup and places the results as appropriate in the output
+queue and modifies the input queue as needed. 
+
+The format of the input/output queues will likely be adjusted or additional
+functor signatures will be provided. 
+
+Both the top level and FusionGroup functors can be overridden by the user. The
+API provides the functor signature and handle but otherwise the API does
+not constrain how the functors  operate.
+
+This gives the necessary generality and extensibility while keeping the
+API definition unobtrusive.
+
+The remainder of the code and support classes are template based. 
+
+## Domain specific language
+
+This is a proposed syntax. Work on this is deferred  in favor of the 
+C++ api.
+
+A domain specific language is specified and an interpreter and compiler are (will be) provided.  The domain specific language expresses sequences, constraints and transformations. The expressions from the DSL can be used in pre-compiled or interpreted form. Once either interpretation or compilation is performed the run time interface to the model is the same C++ API as above.
+
+The DSL has a placeholder definition and a syntactically consistent lexer
+and parser but no action code has been implemented for the grammar. Current
+focus in on the C++ API and the interface to CAM.
+
+```
+
+ 1. fusion fs1 {
+ 2.
+ 3.   isa   rv64g
+ 4.   uarch oly1
+ 5.   input_seq in_seq
+ 6.
+ 7.   sequence seq1(in_seq,rv64g) {
```

Comment:
I wonder if the `rv64g` here is redundant since the fusion defines this on line "3"

Reply:
It may be, rv64g is a reference to the isa decl object on line 162. I've kept the arg list for this version so that when a sequence decl is a top level object it is consistent. This is not well documented.

I think we should have a discussion on the dsl syntax. I threw something out there as a strawman.